### ISSUE TITLE
Decimal fields dropped in group by with more than one column

### ISSUE THREAD

**Issue by drudd** [pandas-dev/pandas#22275]
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import decimal

df = pd.DataFrame({'a': [decimal.Decimal('4.56')]*6, 'b': range(3, 6)*2, 'c': range(6)})
print df.groupby('b')['a'].sum()
print df.groupby('b')['a', 'c'].sum()
print df.groupby('b').agg({'a': 'sum', 'c': 'sum'})
```

Output:
```
b
3    9.12
4    9.12
5    9.12
Name: a, dtype: object
   c
b
3  3
4  5
5  7
      a  c
b
3  9.12  3
4  9.12  5
5  9.12  7
```

#### Problem description

The aggregation over column a is dropped when another field is accessed from the groupby object, but works when requested through `agg` (I'm not sure if 'sum' is exactly equivalent to `.sum()` in the above).

#### Expected Output

```
b
3    9.12
4    9.12
5    9.12
Name: a, dtype: object
      a  c
b
3  9.12  3
4  9.12  5
5  9.12  7
      a  c
b
3  9.12  3
4  9.12  5
5  9.12  7
```

#### Output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.15.final.0
python-bits: 64
OS: Darwin
OS-release: 17.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.23.4
pytest: None
pip: 18.0
setuptools: 39.1.0
Cython: None
numpy: 1.14.3
scipy: 1.1.0
pyarrow: 0.8.0
xarray: None
IPython: 5.7.0
sphinx: None
patsy: 0.5.0
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: 1.0.4
lxml: None
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.7
pymysql: None
psycopg2: 2.7.4 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

**Comment by drudd** [comment]
Tracing through, the 'a' column appears to be dropped by this section of `NDFrameGroupBy. _cython_agg_blocks`:

```
        if numeric_only:
            data = data.get_numeric_data(copy=False)
```
Specifically this line: https://github.com/pandas-dev/pandas/blob/0409521665bd436a10aea7e06336066bf07ff057/pandas/core/groupby/groupby.py#L4006

This suggests to me that there's a mismatch with the logic that sets `numeric_only` (True), and `.get_numeric_data` which skips the Decimal field, encoded as objects.

**Comment by WillAyd** [comment]
There is no native support for Decimal types in NumPy and therefore it is also generally weakly supported in pandas. In theory a community-contributed Extension would help improve that.

In the meantime though I do agree that it's strange that there is a difference between the Series and DataFrame GroupBy mechanism's. Investigation and PRs to align that behavior are certainly welcome

**Comment by drudd** [comment]
My above comment appears to be correct, `numeric_only = True` is set for `sum` when that groupby function is created:  
https://github.com/pandas-dev/pandas/blob/0409521665bd436a10aea7e06336066bf07ff057/pandas/core/groupby/groupby.py#L1466

Which can be seen by using a non-numeric_only group function such as min:
```
df.groupby('b')['a', 'c'].min()
```
produces
```
      a  c
b
3  4.56  0
4  4.56  1
5  4.56  2
```

Not knowing this api very closely I'm surprised that an exception isn't raised earlier in the process if one of the fields is non-numeric (and would be happy if it did so rather than silently dropping data).

**Comment by drudd** [comment]
Another tactic would be to remove the `get_numeric_data` call and rely on the try/except later in the loop which should get raised if there is no appropriate aggregation function.

I have tested this and it restores consistency between the two methods. Happy to PR if we can be confident this won't break other logic.

**Comment by WillAyd** [comment]
If you have something that works and doesn't break other tests locally feel free to submit as a PR

**Comment by drudd** [comment]
Unfortunately it causes a series of test failures, so the solution isn't quite so simple.

**Comment by drudd** [comment]
This bug is similar to #21664 and I think the failing tests may be expecting 
http://pandas.pydata.org/pandas-docs/stable/groupby.html#automatic-exclusion-of-nuisance-columns

**Comment by jschendel** [comment]
> There is no native support for Decimal types in NumPy and therefore it is also generally weakly supported in pandas. In theory a community-contributed Extension would help improve that.

As a side note, we actually have this within our test suite, albeit in a fledgling state:

https://github.com/pandas-dev/pandas/blob/0370740034978d3a63d4b8e5e2c96ff54e7e08ba/pandas/tests/extension/decimal/array.py#L37-L118

This doesn't quite get the job done though, as there's currently no way to dynamically alter `ExtensionBlock.is_numeric`, which currently is always `False`, and is what `get_numeric_data` is ultimately looking at for `DecimalArray`.

Will open a separate issue for the above though, as it's only tangentially related to this issue, and using `DecimalArray` would really only resolve this issue for the `Decimal` case (i.e. a more generic solution for `object` dtype might be nice?).

**Comment by jreback** [comment]
I think this will just work after #22762 (if constructed as a ``DecimalArray``)

**Comment by adisunw** [comment]
Any fix for this? Currently just ran into this same problem. 

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.6.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.1
numpy            : 1.16.4
pytz             : 2019.1
dateutil         : 2.8.0
pip              : 19.0.3
setuptools       : 40.8.0
Cython           : None
pytest           : 5.0.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.3.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.5.0
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.3.3
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : 1.3.4
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

**Comment by rhshadrach** [comment]
This was resolved by #46560
